### PR TITLE
Addition of a force argument to Tensor.numpy()

### PR DESCRIPTION
**Pull request by martinPasen**
solving issues:
- #20778
- #59945

In this PR i have added force argument to Tensor.numpy(). It is set to False by default. If it's set to True then it:
1. detatches Tensor, if requires_grad == True
2. moves to cpu, if it's not on cpu already
3. uses .resolve_conj() if .is_conj() == True
4. [updated] uses .resolve_neg() if .is_neg() == True

### CODE REVIEW

**Comment by vadimkantorov** [comment]
About sparse: but NumPy also supports some sparse formats. Should we consider converting to them?

File: torch/csrc/utils/tensor_numpy.cpp
Comment (albanD) [inline torch/csrc/utils/tensor_numpy.cpp:119]: You can simplify this code by unconditionally doing `auto new_tensor = tensor.detach().cpu().to_dense()` after all the checks. As these ops will be no-ops if the tensor is already requires_grad=False, on cpu and strided.

Also that will give you a simple way to document what the force keywork is doing.

File: torch/csrc/utils/tensor_numpy.cpp
Comment (albanD) [inline torch/csrc/utils/tensor_numpy.cpp:114]: nit: you can use !force instead of force == false

**Review by albanD — COMMENTED** [review]
Thanks for the PR!

Could you update the doc for the function to reflect the new argument and what it does please?

File: torch/csrc/utils/tensor_numpy.cpp
Comment (martinPasen) [reply torch/csrc/utils/tensor_numpy.cpp:119]: When i implement it like this, i get:

`NotImplementedError: Could not run 'aten::to_dense' with arguments from the 'CPU' backend. This could be because the operator doesn't exist for this backend, or was omitted during the selective/custom build process (if using custom build). If you are a Facebook employee using PyTorch on mobile, please visit https://fburl.com/ptmfixes for possible resolutions. 'aten::to_dense' is only available for these backends: [MkldnnCPU, SparseCPU, SparseCsrCPU, BackendSelect, Named, Conjugate, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, UNKNOWN_TENSOR_TYPE_ID, AutogradMLC, AutogradHPU, AutogradNestedTensor, AutogradPrivateUse1, AutogradPrivateUse2, AutogradPrivateUse3, Tracer, UNKNOWN_TENSOR_TYPE_ID, Autocast, Batched, VmapMode].`

Is my build wrong ? Or you just can't call .to_dense() on dense tensor. BTW i expected it to be no-op as well and tried to implement it similarly.

Comment (albanD) [reply torch/csrc/utils/tensor_numpy.cpp:119]: Ho, well my bad. Apparently it is not..
Can you just to do it for detach() and cpu() then please?

Also you can link to this issue for the dense part saying that we should update that code once this is solved: https://github.com/pytorch/pytorch/issues/59958

**Comment by pearu** [comment]
> 3. converts to dense Tensor, if it's in sparse representation

Is this really necessary? On a beautiful day in the future, numpy may support sparse formats as well, and then this feature becomes a BC issue.

**Comment by vadimkantorov** [comment]
Maybe extra `dense=True` argument? Another option is to take a lazy dependency on SciPy and return scipy.sparse.coo_matrix, scipy.sparse.csr_matrix when possible (I think it's now limited for ndim=2)

**Comment by martinPasen** [comment]
>On a beautiful day in the future, numpy may support sparse formats as well, and then this feature becomes a BC issue.
Makes sense. Should I implement it now, or is there someone that is supposed to have the final word?

**Comment by mruberry** [comment]
Make sure to update the docs, make the "force" argument kwarg-only, and just have it do the equivalent of `t.detach().cpu()'. I suggest not bothering with sparse tensors.

**Comment by martinPasen** [comment]
> Make sure to update the docs, make the "force" argument kwarg-only, and just have it do the equivalent of `t.detach().cpu()'. I suggest not bothering with sparse tensors, but handling the conjugate bit seems OK.
> 
> Edit: Sorry -- wasn't totally clear. Also, I think `force` should only make a copy if necessary. So if a CPU tensor can share memory with the NumPy tensor (like by calling `.numpy()` on it today) we probably want to preserve that behavior. If we want an argument that always makes a copy then we should consider a different arg name, like `copy`.

Makes sense. The following code satisfies that right ?
`auto prepared_tensor = tensor.detach().cpu().resolve_conj();`

`.cpu()` and `resolve_conj()` are no-ops if the tensor is already on cpu or conjugate bit is False. `detach()` is an in-place operation.  `.cpu()` makes copy only if tensors device is not cpu and in that situation it's necessary. `resolve_conj()` also makes copy, if conjugate bit is set to True. Is that what we want ? Or would it be better to have in-place version of `resolve_conj()` ? Does in-place version exist ?

**Comment by martinPasen** [comment]
@soulitzer

File: torch/_tensor_docs.py
Comment (albanD) [inline torch/_tensor_docs.py:2648]: You want to specify here that this is kwarg only?

File: torch/_tensor_docs.py
Comment (albanD) [inline torch/_tensor_docs.py:2656]: sparse

File: torch/_tensor_docs.py
Comment (albanD) [inline torch/_tensor_docs.py:2659]: It always succeeds no? So the "tries" can be removed

File: torch/_tensor_docs.py
Comment (albanD) [inline torch/_tensor_docs.py:2661]: We always call resolve_conj right?

File: torch/csrc/utils/tensor_numpy.cpp
Comment (albanD) [inline torch/csrc/utils/tensor_numpy.cpp:136]: You can inconditionally resolve conj no?

**Review by albanD — COMMENTED** [review]
Thanks for the updates it looks almost ready.

File: torch/_tensor_docs.py
Comment (martinPasen) [reply torch/_tensor_docs.py:2648]: probably yes, but not sure how. Like this?
`numpy(*, force=False) -> numpy.ndarray`

File: torch/_tensor_docs.py
Comment (martinPasen) [reply torch/_tensor_docs.py:2659]: No, it throws an exception if the tensor is sparse. Probably should explicitly say that.

File: torch/_tensor_docs.py
Comment (martinPasen) [reply torch/_tensor_docs.py:2661]: Yes, just realized it's a no-op as well if is_conj() == False. Will fix that in docs as well as in the code.

**Comment by martinPasen** [comment]
> Thanks for the updates it looks almost ready.

what about the build/test fails? They seem pretty "random" - they feel to be not connected to my code. Should i "just look deeper", or should I ignore them? 

One more question. 2 reviewers mean that both of you have to review, or one is enough?

**Comment by vadimkantorov** [comment]
Does it also work for MKL_DNN tensors?

File: torch/_tensor_docs.py
Comment (albanD) [reply torch/_tensor_docs.py:2648]: Yes

**Comment by albanD** [comment]
> Does it also work for MKL_DNN tensors?

The `.cpu()` would take care of that no?

File: torch/_tensor_docs.py
Comment (albanD) [inline torch/_tensor_docs.py:2665]: nit: extra "," in the middle of the sentence. Also "would otherwise throw" sounds a little better.

**Review by albanD — APPROVED** [review]
Looks good. a minor update to the phrasing and this will be good to go.

No need to get a review from every reviewer, one is enough

File: torch/_tensor_docs.py
Comment (albanD) [reply torch/_tensor_docs.py]: What I meant is remove all the commas here :p

File: test/test_numpy_interop.py
Comment (mruberry) [inline test/test_numpy_interop.py:175]: Shouldn't the lambda here be `x.numpy(force=True)`?

File: torch/_tensor_docs.py
Comment (mruberry) [inline torch/_tensor_docs.py:2661]: "If tensors storage" -> "If the tensor's storage"

File: test/test_numpy_interop.py
Comment (albanD) [reply test/test_numpy_interop.py:175]: The one I prefer by the way is:
```
with self.assertRaisesRegex(TypeError, "part of the error message"):
    x.numpy(force=True)
```

This makes sure that the error that is raised is the one you expect and not another unrelated TypeError.

File: torch/_tensor_docs.py
Comment (martinPasen) [reply torch/_tensor_docs.py:2661]: > The tensor and the returned ndarray will only share their storage if the conversion would have happened with :attr:`force` being False.

I don't think that's true. If we have tensor that is:

- dense
- on CPU
- not conjugated
- requires grad

then t.numpy() throws an exception, but t.numpy(force = True) returns ndarray that shares its storage.

Comment (pearu) [reply torch/_tensor_docs.py]: ".. if the conversion would have happened with :attr:`force` being False and the tensor being detached" ?

Comment (martinPasen) [reply torch/_tensor_docs.py]: In the last commit i went with:

If the tensor
isn't on the CPU or its conjugate bit is set, the tensor won't share
its storage with the returned ndarray.

File: test/test_numpy_interop.py
Comment (martinPasen) [reply test/test_numpy_interop.py]: in last 2 commits i have unified error messages you can get in tensor.numpy(). CI failed, because some test used `self.assertRaisesRegex()`. Is my change breaking BC ? Should i return to the previous eroor messages ?

File: torch/_tensor_docs.py
Comment (mruberry) [reply torch/_tensor_docs.py]: I think it's an interesting UX question whether a tensor with requires_grad=True should share its storage with the NumPy array created or not. I kinda lean towards creating a copy in that case for clarity and to avoid troublesome scenarios later.

What do you think, @albanD?

Comment (pearu) [reply torch/_tensor_docs.py]: In the case of `requires_grad=True` (and if other view conditions are met), I suggest returning a read-only numpy array (set [writable flag](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.setflags.html#numpy-ndarray-setflags) to False) that shares storage with the tensor in order to avoid copies.

Comment (martinPasen) [reply torch/_tensor_docs.py]: If I look back to the issues and all previous communication, I think, that the purposes of the force argument are

- to be a shorthand for `tensor.detach().resolve_conj().cpu().to_dense().numpy()`
- to give make it easier to work with many libraries at once, so if you are trying to create some visualization, you don't care if you  have a Torch tensor, or a TensorFlow tensor or a Panda array or whatever ... you just call numpy(force=True)

We have already "broken" that goal by letting sparse tensors throw an exception insight of potential future BC problems. I don't think we should break it again if we don't have a serious reason. And I don't see any reason.

Comment (albanD) [reply torch/_tensor_docs.py]: I think this is ok to have a regular numpy array that shares data.
It is just like when you do `.detach()` in general, any inplace change is not differentiable and won't be tracked. You get the same behavior here.

Comment (albanD) [reply torch/_tensor_docs.py]: And I don't think we want to diverge from the `tensor.detach().resolve_conj().cpu().numpy()` definition.

File: test/test_numpy_interop.py
Comment (albanD) [reply test/test_numpy_interop.py]: It is ok to update the error message. We don't consider this BC in general.
You can just update the tests to make them pass again.

File: torch/_tensor_docs.py
Comment (pearu) [reply torch/_tensor_docs.py]: For what it's worth, `to_dense()` is not always realizable, for instance, when a sparse tensor has many large dimensions and one may not have enough RAM to store such a tensor using a strided layout.

Different from a conjugate bit, a sparse layout is not a lazy variant of a dense tensor. The result of `to_dense()` operation may depend on how are the unspecified elements interpreted. For instance, are they zeros or do sparse tensors represent masked tensors? In the latter case, one may want to get a masked numpy array as a result of `.numpy()`. If `numpy(force=True)` will involve `to_dense`, one has to fix the semantics of sparse tensors. I am not sure that this PR, that main objective is to introduce a shorthand to a convenience transformation, is the proper place to make these decisions.

Comment (albanD) [reply torch/_tensor_docs.py]: Notice that I didn't put `to_dense()` in the snippet above ;) 
I agree with you that it is a much trickier business and we can leave that to a later PR.

Comment (pearu) [reply torch/_tensor_docs.py]: Yes, @albanD , I noticed it but only after pressing the Comment button ;)

File: test/test_numpy_interop.py
Comment (mruberry) [inline test/test_numpy_interop.py:157]: Style nit:

```
for force, requires_grad, sparse, conj in itertools.product(*(((False, True),) * 4)):
```

will save you so much indenting

Comment (mruberry) [reply test/test_numpy_interop.py:157]: That said, I think it might make more sense to just create smaller, separate tests for these properties: requires_grad, sparse, and conj.

**Review by mruberry — APPROVED** [review]
LGTM, @martinPasen! Thanks for the PR.

I think the test could be simplified (and I made a couple inline comments, mostly I think it could just be separated into much simpler tests), but I don't think that's a blocking issue.

**Comment by mruberry** [comment]
@albanD are you merging this?

**Comment by albanD** [comment]
Yep

File: test/test_numpy_interop.py
Comment (martinPasen) [reply test/test_numpy_interop.py:157]: I wasn't really sure how to divide them into smaller tests. My goal was to test every possible case out of 2^4 options. Do you have an idea how to separate them?

**Comment by martinPasen** [comment]
> @albanD are you merging this?

> Yep

I am really thankful for your patience and help. This is my first real PR and it probably takes longer than it should, but I learned a lot, thanks for that. I have some more questions if that's ok.

1. what does it mean, that PR was imported? Does it mean, that it's in a "prepared state" so it's sent for broader tests?
2. again there are some errors and skipped tests. Most of them seem to not have a connection to the PR, but there are two that probably should be fixed.

The first is in a test. I used 
` error_msg = "Use tensor\..*\.numpy\(\) instead\."`
what fails the flake8-py3 test and should be probably fixed to
` error_msg = r'Use tensor\..*\.numpy\(\) instead\.'`

The second is in documentation, line 2712, which is too long. Flake8-py3 again throws an error. When I tried to fix it by just "pressing enter" the formatting broke and it was interpreted as another argument. What is the best practice in the situation? Am I supposed to shorten the line, or is there a way to split it into two lines without being interpreted as another argument?

Side note - I won't be online in the following 2 weeks. If the PR is in a hurry, then it would be nice if someone else finished the PR, if not I am looking forward to finishing it when I come back.

**Comment by albanD** [comment]
Hi,

> what does it mean, that PR was imported? Does it mean, that it's in a "prepared state" so it's sent for broader tests?

That means it was imported into the fb internal code base for extra testing. I am taking care of that, so no need to worry about that on your end.

> again there are some errors and skipped tests. Most of them seem to not have a connection to the PR, but there are two that probably should be fixed.

Yes these tests should be fixed. Breaking the line whichever way you can should work here.
I'll do the update/land if I have some time later this week.

**Comment by mruberry** [comment]
ping @albanD

**Comment by albanD** [comment]
I think we would still be happy to merge that. This needs rebasing and fixing CI though.

**Comment by HaoZeke** [comment]
I'm happy to update this if there are no other takers in a few days :)

**Comment by HaoZeke** [comment]
I think this should be alright now. However, since this was partially imported I'm not sure the GH-First workflow is appropriate. Hence the review requests.

**Comment by albanD** [comment]
I think creating a new PR might be simpler to avoid having to deal with the partial import. Do you think you could open a new one and add me as a reviewer there?

**Comment by HaoZeke** [comment]
> I think creating a new PR might be simpler to avoid having to deal with the partial import. Do you think you could open a new one and add me as a reviewer there?

Makes sense, done in #78564

**Comment by albanD** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/pull/78564

**Pull request closed without merging**
